Misc Cortex-R changes

bbolen opened the pull request:
This cleans up a couple of minor nits.

There is also a patch that dumps out the callee saved registers on a fatal exception.  These registers can be very helpful when debugging crashes.

example of output
```
*** Booting Zephyr OS build zephyr-v2.4.0-4934-gc2a5cf2acb85  ***
Coredump: qemu_cortex_r5
E: ***** UNDEFINED INSTRUCTION ABORT *****
E:  r4: 0x00001519  r5:  0x00000000  r6:  00000000
E:  r7: 0x00000000  r8:  0x00000000  r9:  00000000
E: r10: 0x00000000 r11:  0x00000000 psp:  00003c48
E: r0/a1:  0x00000019  r1/a2:  0x0000000a  r2/a3:  0xff00002c
E: r3/a4:  0x00000000 r12/ip:  0x0000000a r14/lr:  0x0000019b
E:  xpsr:  0x4000013f
E: Faulting instruction address (r15/pc): 0x0000019e
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x2e78 (unknown)
E: Halting system
```

Commit:
arch: arm: aarch32: Use ARRAY_SIZE in for loop

Do not hardcode the array size in the loop for printing out the floating
point registers of the exception stack frame.  The size of this array
will change when Cortex-R support is added.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: aarch32: cortex_a_r: Correct syntax for srs

The writeback specification should be after the register, not after the
mode according to the documentation at

Link: https://developer.arm.com/documentation/dui0489/h/arm-and-thumb-instructions/srs

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

stephanosio reviewed (commented):
Overall the changes look ok to me.

For the callee-saved register dumping part, the Cortex-M implementation only does that when `CONFIG_EXTRA_EXCEPTION_INFO` is enabled, so I think it would be nice to have the same behaviour for the Cortex-A/-R.

bbolen commented:
@stephanosio I agree that would be preferable, but I think I need to talk this through.  Enabling CONFIG_EXTRA_EXCEPTION_INFO enables printing out the callee saved registers in arch/arm/core/aarch32/fatal.c, which would get rid of a function I added.  However, esf_dump already does this by dereferencing the esf pointer, meaning the exception code in exc.S needs to decrement the stack pointer by the size of struct __esf and save all the extra registers there.

At this point things get thorny.  The exception entry/exit code uses srsdb/rfeia.  If we decrement $sp by the size of __esf on exception entry, on the exception return we will leak stack space since the return state is now in the middle of the stack frame and not at the top.  In other words, the exception return code expects the return address and cpsr to be at the top of the stack.

I see 3 options.
a) On exception entry, save the stack frame as normal.  Then subtract off the size of __esf and copy the previously saved stack frame to the new location and call the fault handling code with that.
b) On exception entry, subtract off the size of __esf then save the stack frame.  Then before calling z_arm_exc_exit, move the return state (return lr and sp) up the stack to where z_arm_exc_exit expects it.
c) Split struct __basic_sf into 2 structs, one with r0-r4,ip,lr and the other with pc,xpsr.  The second struct would move to the end of struct __esf.  We could move the entire basic_sf to the end of __esf for cortex_a_r as well.

Thoughts?

Commit:
arch: arm: aarch32: cortex_a_r: Dump callee saved registers on fault

Some of these registers may contain nuggets of information that would be
beneficial when debugging, so include them in the fault dump.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

carlescufi commented:
@ioannisg @carlocaione can you take a look please

carlescufi commented:
@npitre @povergoing same as above, a quick review would be appreciated

povergoing commented:
> On exception entry, subtract off the size of __esf then save the stack frame. Then before calling z_arm_exc_exit, move the return

I am not sure, is 1st option the simplest way to solve this problem? The second one seems to cause the code hard to read? But anyway, it's my personal opinion.

bbolen commented:
> > On exception entry, subtract off the size of __esf then save the stack frame. Then before calling z_arm_exc_exit, move the return
> 
> I am not sure, is 1st option the simplest way to solve this problem? The second one seems to cause the code hard to read? But anyway, it's my personal opinion.

Yes, the second option was confusing.  The first option ended up being simple and seems the easiest to read, just save the esf for the exception return, subtract the stack pointer, and save the esf again for the dump of the exception data.  I went with that one.

povergoing commented on arch/arm/core/aarch32/cortex_a_r/exc.S:
```diff
@@ -61,7 +61,21 @@ SECTION_SUBSEC_FUNC(TEXT, __exc, z_arm_undef_instruction)
-	srsdb sp, #MODE_UND!
+	srsdb sp!, #MODE_UND
+	stmfd sp, {r0-r3, r12, lr}^
+	sub sp, #24
```
Not sure will it be better to pack these similar codes into a macro or something in the future?  Anyway, LGTM.
My only concern maybe is (not sure because it's only related to exceptions, abort, etc.), since it will always cost double the space of the stack, will it cause some kind of performance problems?

stephanosio replied:
It looks like the undef exception could potentially be problematic in the future since it could be used for emulating instructions; but, IMO, that that's something for a future PR that actually implements such features.

carlescufi commented:
@stephanosio any chance you could take a look at this PR?

stephanosio reviewed (approved):
I would have preferred to see the esf-related structs reworked to better suit the non-Cortex-M architectures such that duplication of the "basic sf" in the stack is not necessary, but that would require a substantial amount of work.

Noting that exception handling is not a regular occurrence and the performance impact is minimal, this should be good enough for now. In the future, this issue can be revisited when the majority of the arch code for the M and the rest of ARM profiles are split (which would become inevitable as we add more features).
